lib/srv/desktop/rdp/rdpclient/src/rdpdr/mod.rs

```diff
@@ -350,75 +833,690 @@ impl Client {
-        Ok(Self {
-            component: Component::from_u16(component).ok_or_else(|| {
-                invalid_data_error(&format!("invalid component value {:#06x}", component))
-            })?,
-            packet_id: PacketId::from_u16(packet_id).ok_or_else(|| {
-                invalid_data_error(&format!("invalid packet_id value {:#06x}", packet_id))
-            })?,
-        })
+
+    pub fn handle_tdp_sd_write_response<S: Read + Write>(
+        &mut self,
+        res: SharedDirectoryWriteResponse,
+        mcs: &mut mcs::Client<S>,
+    ) -> RdpResult<()> {
+        debug!("received TDP: {:?}", res);
+        if let Some(tdp_resp_handler) = self
+            .pending_sd_write_resp_handlers
+            .remove(&res.completion_id)
+        {
+            let rdp_responses = tdp_resp_handler(self, res)?;
+            let chan = &CHANNEL_NAME.to_string();
+            for resp in rdp_responses {
+                mcs.write(chan, resp)?;
+            }
+            Ok(())
+        } else {
+            return Err(try_error(&format!(
+                "received invalid completion id: {}",
+                res.completion_id
+            )));
+        }
     }
-    fn encode(&self) -> RdpResult<Vec<u8>> {
-        let mut w = vec![];
-        w.write_u16::<LittleEndian>(self.component.to_u16().unwrap())?;
-        w.write_u16::<LittleEndian>(self.packet_id.to_u16().unwrap())?;
-        Ok(w)
+
+    fn prep_device_create_response(
+        &mut self,
+        req: &DeviceCreateRequest,
+        io_status: NTSTATUS,
+        new_file_id: u32,
+    ) -> RdpResult<Vec<Vec<u8>>> {
+        let resp = DeviceCreateResponse::new(req, io_status, new_file_id);
+        debug!("sending RDP: {:?}", resp);
+        let resp = self
+            .add_headers_and_chunkify(PacketId::PAKID_CORE_DEVICE_IOCOMPLETION, resp.encode()?)?;
+        Ok(resp)
     }
-}
+
+    fn prep_query_info_response(
+        &self,
+        req: &ServerDriveQueryInformationRequest,
+        file: Option<&FileCacheObject>,
+        io_status: NTSTATUS,
+    ) -> RdpResult<Vec<Vec<u8>>> {
+        let resp = ClientDriveQueryInformationResponse::new(req, file, io_status)?;
+        debug!("sending RDP: {:?}", resp);
+        let resp = self
+            .add_headers_and_chunkify(PacketId::PAKID_CORE_DEVICE_IOCOMPLETION, resp.encode()?)?;
+        Ok(resp)
+    }
+
+    fn prep_device_close_response(
+        &self,
+        req: DeviceCloseRequest,
+        io_status: NTSTATUS,
+    ) -> RdpResult<Vec<Vec<u8>>> {
+        let resp = DeviceCloseResponse::new(req, io_status);
+        debug!("sending RDP: {:?}", resp);
+        let resp = self
+            .add_headers_and_chunkify(PacketId::PAKID_CORE_DEVICE_IOCOMPLETION, resp.encode()?)?;
+        Ok(resp)
+    }
+
+    fn prep_drive_query_dir_response(
+        &self,
+        device_io_request: &DeviceIoRequest,
+        io_status: NTSTATUS,
+        buffer: Option<FileInformationClass>,
```

Comment:
Why is this called buffer?

Reply:
It's an RDP structure and that's what RDP calls it: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/9c929407-a833-4893-8f20-90c984756140